### DISCUSSION TITLE
ImGui integration for external project

### DISCUSSION THREAD

**Discussion by yunusyurtturk** [pelicanmapping/osgearth#2621] [Support Q&A] [answered]
Hi;
I played enough with osgEarth library itself and now I'm preparing my external project.  My project will not be under/osgEarth library, instead it will be in its own path & will link to osgEarth. 

For osgEarth integration part, I'm able to link & compile my external project. Now I need to integrate ImGui to my external project. In this step, I'm facing some issues:

When I check the CMakeLists.txt file of the osgearth_imgui project, it contains this code, which is defined in osgearth-macros.cmake file:

```
add_osgearth_app(
    TARGET osgearth_imgui
    SOURCES osgearth_imgui.cpp
    USE_IMGUI
    FOLDER Tools )
```

I copied this helper cmake to my project and myCMakeLists.txt is (there are additional SETs, which are not shown here): 

```
add_osgearth_app(
        TARGET ${PROJECT_NAME}
        SOURCES main.cpp
        USE_IMGUI
        INCLUDE_DIRECTORIES ${OSGEARTH_SOURCE_DIR}/src/osgEarth/ImGui ) # !Attention to this part
```

The reason I want to use ImGui code inside osgearth/src/ImGui is there are predefined integration code and some UI in those files. I also added ImGui as a submodule to my project (as it is done in osgEarth) so that build process continues smoothly with ImGui.

**Problem 1:**
${OSGEARTH_SOURCE_DIR}/src/osgEarth/ImGui/ImGui.cpp => This line is used in add_osgearth_app() in osgearth-macros.cmake. This is a source library, which makes me unable to use just 'INSTALL' directory for integration purposes, there have to be source code available.

**Problem 2:**
My main.cpp include order is:

```
#include <osgEarth/ImGui/ImGuiApp>
#include <osgEarth/MapNode>
#include <osgEarth/Map>
...
```

 I'm unable to compile with this order because I'm getting `MAX_PATH` and `GetLogicalDriveStringsA` undefined error.

So I added `windows.h` to to my includes:

```
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <osgEarth/ImGui/ImGuiApp>
#include <osgEarth/MapNode>
#include <osgEarth/Map>
```

However, this time I'm getting a lot of error in StringUtils, Units, weejobs etc. headers, symbols/macros are conflicting.
I have to place windows.h before ImGuiApp include line, also I need to include ImGuiAPp at topmost places because it need to be included before gl.h (GLEW issues)

This time my problem is, am I doing right to trying to use "osgEarth/ImGui/ImGuiApp"? Is my method correct or is it wrong? If wrong, what can you suggest?

**Comment by yunusyurtturk** [comment]
Additionally, when used  with `WIN32_LEAN_AND_MEAN `and `NOMINMAX`

```
#define WIN32_LEAN_AND_MEAN
#define NOMINMAX
#include <windows.h>
#include <osgEarth/ImGui/ImGuiApp>
...
```

I'm still getting MAX_PATH and GetLogicalDriveStringsA not defined errors. `#define NOMINMAX` resolves conflicts I mentioned above, but I'm still getting not defined errors with this config.

**Comment by gwaldron** [comment, marked as answer]
The master branch of osgEarth has a share/ExternalProject folder which demonstrates how to do this properly. You do not need to copy any of the macros into your own project. Give that a try and see how you get along. Later this month we will release a dot version that incorporates the new correct cmake setup.
